### PR TITLE
ENH: Rename resources deterministically in merge_page

### LINKED ISSUE #1532: merge_page isn't reproducible due to resource renaming
Similar to https://github.com/py-pdf/pypdf/issues/1531, it's nice if code is reproducible, meaning same inputs gives exactly the same outputs, e.g. to be able to reproduce hashes or to do snapshot testing.

Currently it seems like using merge_page on two pages that have identically-named resources with different contents _always_ rename them using a random UUID. It'd be nice to either:

- use some other scheme for choosing a name that _is_ reproducible
- allow users to provide a custom function so they can make it reproducible

https://github.com/py-pdf/pypdf/blob/ea598dda99c6823d637e28f2f558ee5877499146/pypdf/_page.py#L585

Thank you for pypdf!

## Environment

Which environment were you using when you encountered the problem?

```bash
$ python -m platform
macOS-12.5.1-arm64-arm-64bit

$ python -c "import pypdf;print(pypdf.__version__)"
3.2.0
```

## Code + PDF

This is a minimal, complete example that shows the issue:

```python
import io
from pypdf.generic import NameObject, ArrayObject, DictionaryObject
from pypdf import PdfWriter, PageObject

def write() -> bytes:
    pdf = PdfWriter()

    page1 = PageObject.create_blank_page(width=100, height=100)
    page2 = PageObject.create_blank_page(width=100, height=100)

    # BUG
    page1[NameObject("/Resources")][NameObject("/Properties")] = DictionaryObject({NameObject("/A"): NameObject("/1")})
    page2[NameObject("/Resources")][NameObject("/Properties")] = DictionaryObject({NameObject("/A"): NameObject("/2")})
    page1.merge_page(page2)

    pdf.add_page(page1)

    buf = io.BytesIO()
    pdf.write(buf)
    return buf.getvalue()

print(write().decode("utf-8", errors="backslashreplace"))
```

Running that script like `diff <(python test.py) <(python test.py)` shows that the output differs between runs, e.g.

```diff
@@ -25,5 +25,5 @@
 /Properties <<
 /A /1
-/A0368617b-e584-4d61-bcc2-fb66eadf2459 /2
+/A3f1b94aa-e8c1-4835-9140-ee61b9471ebf /2
 >>
 /ProcSet [ ]
```

## Traceback

This is the complete Traceback I see: N/A

### PR DESCRIPTION
**Pull request by huonw**
This fixes #1532 by adjusting the procedure used for renaming resources in `merge_page`, so that resources that have the same name (but different contents) are renamed in a deterministic/reproducible way.

When merging pages that both have a resource of the same name (say, `/Example`) but different values, the resource from the second page will now be renamed to `/Example-0`. Previously, it would be renamed with a random UUIDv4 (e.g. `/Exampledbbbe7cb-5f34-4061-b863-41919b326b49`) which would be different on every run, even if the inputs were identical.

The renamed name may already exist if a PDF is carefully/maliciously crafted, in which case `/Example-1`, `/Example-2`, etc. are tried until an appropriate name is found. If any of these options have the same value, that name/value is reused.

The `/Example-0` pattern is short and sweet, but can easily be changed, e.g. some options off the top of my head:

- `/Example-pypdf-merged-0` for explanation about where it comes from
- `/Example-9c4beb01-0` where the `9c4beb01` is some marker hardcoded into pypdf's code, to reduce the chance of a "normal" PDF requiring `idx > 0` (only truly maliciously crafted ones)

### CODE REVIEW

File: pypdf/_page.py [hunk 1]
```diff
@@ -576,17 +575,41 @@ def _merge_resources(
...
+            # many keys, thus this'll eventually end, even if it's been crafted
+            # to be maximally annoying.
+            computed_key = base_key
+            idx = 0
+            while computed_key in new_res:
+                if new_res.raw_get(computed_key) == value:
+                    # there's already a resource of this name, with the exact
+                    # same value
+                    return computed_key, True
+                computed_key = "%s-%d" % (key, idx)
```
Comment (MasterOdin) [inline pypdf/_page.py:594]: Use [f-strings](https://realpython.com/python-f-strings/) as they're much faster than [other string formatting methods.](https://www.scivision.dev/python-f-string-speed/).

```suggestion
                computed_key = f"{key}-{idx}"
```

Comment (huonw) [reply pypdf/_page.py:594]: Ah, sorry, I thought they weren't supported in some versions of Python supported by pypdf. I take it pypdf only supports 3.6+?

Comment (MasterOdin) [reply pypdf/_page.py]: Yup:

https://github.com/py-pdf/pypdf/blob/e7e4ffc7e74fd3f1a191bc63527bee0d7986be1f/pyproject.toml#L13

**Comment by MartinThoma** [comment]
Thank you @huonw for this PR and for adding good tests :pray: It's now in `main` and will be released to PyPI today :-)

If you want, I can add you to https://pypdf.readthedocs.io/en/latest/meta/CONTRIBUTORS.html

**Comment by huonw** [comment]
Feel free to add me, thanks 😄 

I've upgraded to 3.3.0 in our code, it works well. Thanks for merging and releasing!

**Pull request merged**